src/services/route.rs

```diff
@@ -0,0 +1,257 @@
+        /// In the event that url crate can't parse the route string, the route string will be passed back to the crate user to use.
+        route: String
+    },
+    /// If the full Url can't be parsed this will be returned
+    CouldNotParseUrl {
+        /// This will contain the full url, not just the route.
+        full_url: String
+    },
+    /// An error indicating that the string passed to the `RouteInfo::parse()` function did not start with a slash.
+    RouteDoesNotStartWithSlash,
+    /// An error indicating that the string passed to the `RouteInfo::parse()` function did not contain ary characters
+    RouteIsEmpty,
+    /// Indicates that the url could not be retrieved from the Location API.
+    CouldNotGetLocationHref
+}
+
+
+impl RouteInfo {
+    /// This expects a string with a leading slash`
+    pub fn parse(route_string: &str) -> Result<RouteInfo, RoutingError> {
+        // Perform some validation on the string before parsing it.
+        if let Some(first_character) = route_string.chars().next() {
+            if first_character != '/' {
+                eprintln!("does not start with slash: '{}'", route_string);
+                return Err(RoutingError::RouteDoesNotStartWithSlash)
+            }
+        } else {
+            return Err(RoutingError::RouteIsEmpty)
+        }
+
+        let full_url = format!("http://dummy_url.com{}", route_string);
+        Url::parse(&full_url)
+            .map(RouteInfo::from)
+            .map_err(|_| RoutingError::CouldNotParseRoute { route: route_string.to_string() })
+    }
+
+    /// Converts the RouteInfo into a string that can be matched upon,
+    /// as well as stored in the History Api.
+    pub fn to_string(&self) -> String {
+        let path = self.path_segments.join("/");
+        let mut path = format!("/{}", path); // add the leading '/'
+        if let Some(ref query) = self.query {
+            path = format!("{}?{}", path, query);
+        }
+        if let Some(ref fragment) = self.fragment {
+            path = format!("{}#{}", path, fragment)
+        }
+        path
+    }
+
+    /// Gets the path segment at the specified index.
+    pub fn get_segment_at_index<'a>(&'a self, index: usize) -> Option<&'a str> {
+        self.path_segments.get(index).map(String::as_str)
+    }
+}
+
+impl From<Url> for RouteInfo {
+    fn from(url: Url) -> RouteInfo {
+        RouteInfo {
+            path_segments: url
+                .path_segments()
+                .expect("The route should always start with '/', so this should never error.")
+                .map(str::to_string)
+                .collect::<Vec<String>>(),
+            query: url.query().map(str::to_string),
+            fragment: url.fragment().map(str::to_string)
+        }
+    }
+}
+
+
+impl RouteService {
+
+    /// Creates a new route service
+    pub fn new() -> RouteService {
+        RouteService {
+            history: window().history(),
+            location: window().location().unwrap(),
```

Comment:
I agree 100% that we should `unwrap` here.

However, I think we should resolve the attributes by unwrapping them as well into a plain struct (panic for `SecurityError`).